packages/gatsby-image/src/index.js

```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
   return convertedProps
 }
 
+/**
+ * Checks if fluid or fixed are art-direction arrays.
+ * @param props   {object}   The props to check for images.
+ * @param prop    {string}   Check for fluid or fixed.
+ * @return {boolean}
+ */
+const hasArtDirectionSupport = (props, prop) =>
+  props[prop] &&
+  Array.isArray(props[prop]) &&
+  props[prop].some(image => typeof image.media !== `undefined`)
+
+/**
+ * Checks for fluid or fixed Art direction support.
+ * @param props   {object}   The props to check for art-direction support.
+ * @return {boolean}
+ */
+const hasArtDirectionArray = props =>
+  hasArtDirectionSupport(props, `fluid`) ||
+  hasArtDirectionSupport(props, `fixed`)
+
+/**
+ * Tries to detect if a media query matches the current viewport.
+ * @param media   {string}  A media query string.
+ * @return {*|boolean}
+ */
+const matchesMedia = ({ media }) =>
+  media ? isBrowser && window.matchMedia(media).matches : false
+
 /**
  * Find the source of an image to use as a key in the image cache.
  * Use `the first image in either `fixed` or `fluid`
  * @param {{fluid: {src: string}[], fixed: {src: string}[]}} args
  * @return {string}
  */
 const getImageSrcKey = ({ fluid, fixed }) => {
-  const data = (fluid && fluid[0]) || (fixed && fixed[0])
+  const data = getCurrentSrcData({ fluid, fixed })
 
   return data.src
 }
 
+/**
+ * Returns the current src - if possible with art-direction support.
+ * @param fluid         {object}    Fluid Image (Array) if existent.
+ * @param fixed         {object}    Fixed Image (Array) if existent.
```

Comment:
Please add the object typing here. You can copy it from the `getImageSrcKey` jsdoc.

Additionally, note that these are properties, not params. However, as suggested below, the param would be one object that is either the fixed or fluid. You can copy the type from the `getImageSrcKey` jsdoc.